tools/testsys/src/wait.rs

```diff
@@ -0,0 +1,359 @@
+use crate::error::{self, Result};
+use futures::future::join_all;
+use log::{error, info, warn};
+use snafu::{OptionExt, ResultExt};
+use std::collections::HashMap;
+use std::time::Duration;
+use testsys_model::clients::{AllowNotFound, CrdClient};
+use testsys_model::test_manager::TestManager;
+use testsys_model::{
+    Crd, CrdExt, DestructionPolicy, Outcome, Resource, ResourceError, TaskState, Test, TestResults,
+};
+use tokio_retry::strategy::{jitter, ExponentialBackoff};
+use tokio_retry::Retry;
+
+const WAIT_TIMEOUT: Duration = Duration::from_secs(600);
+const WAIT_CHECK: Duration = Duration::from_secs(10);
+
+const RETRY_ATTEMPTS: usize = 5;
+
+/// Wait until all of the created crds have completed and return a HashMap containing each `Test`'s
+/// results.
+pub(crate) async fn wait_for_crds(
+    client: &TestManager,
+    crds: &Vec<Crd>,
+) -> Result<HashMap<String, Vec<TestResults>>> {
+    // Get list of all dependencies of the created crds
+    let dependencies: Vec<_> = crds
+        .iter()
+        .filter_map(|crd| match crd {
+            Crd::Test(test) => Some(test.spec.resources.to_owned()),
+            Crd::Resource(resource) => resource.spec.depends_on.to_owned(),
+        })
+        .flatten()
+        .collect();
+
+    // Get a list of all conflicting resources with the created crds
+    let conflicting_resource_names: Vec<_> = crds
+        .iter()
+        .filter_map(|crd| match crd {
+            Crd::Test(_) => None,
+            Crd::Resource(resource) => resource.spec.conflicts_with.to_owned(),
+        })
+        .flatten()
+        .collect();
+
+    // Loop until all conflicting resources have been cleaned up.
+    wait_for_conflicting_resources(client, &conflicting_resource_names).await?;
+
+    // Follow the creation progress of all dependencies.
+    for dependency in dependencies {
+        wait_for_resource_creation(client, &dependency)
+            .await?
+            .context()?;
+    }
+
+    // For each CRD created, add a section to results with the test results.
+    let mut results = HashMap::new();
+    for crd in crds {
+        results.extend(wait_for_crd(client, crd).await?.context()?)
+    }
+
+    Ok(results)
+}
+
+/// Wait until the conflicting resources are deleted.
```

Comment:
Can you define what it means for a resource to be conflicting, here, or somewhere? Maybe it exists somewhere already. I think it means a resource that can't be used by more than one test simultaneously? Or one that cannot be created until another is destroyed?

Reply:
It is described when conflicting resources are identified. Conflicting resources are resources that must not exist before the resource creation can begin.